File: Problems/Insert Delete GetRandom O(1).js

var RandomizedSet = function() {
    this.values = [];
    this.valuesIdx = {};
};

/** 
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.insert = function(val) {
    if (val in this.valuesIdx) {
        return false;
    }

    this.valuesIdx[val] = this.values.length;
    this.values.push(val);

    return true;
};

/** 
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.remove = function(val) {
    if (!(val in this.valuesIdx)) {
        return false;
    }

    const index = this.valuesIdx[val];
    this.valuesIdx[this.values[this.values.length - 1]] = index;
    delete this.valuesIdx[val];
    this.values[index] = this.values[this.values.length - 1];
    this.values.pop();

    return true;
};

/**
 * @return {number}
 */
RandomizedSet.prototype.getRandom = function() {
    const index = Math.floor(Math.random() * this.values.length);
    return this.values[index];
};

File: Problems/Insert Delete GetRandom O(1).py
class RandomizedSet:
    def __init__(self):
        self.values = []
        self.val_idx = {}
    def insert(self, val: int) -> bool:
        if val in self.val_idx:
            return False
        
        self.val_idx[val] = len(self.values)
        self.values.append(val)
        return True
    def remove(self, val: int) -> bool:
        if val not in self.val_idx:
            return False

        del_idx = self.val_idx[val]
        self.val_idx[self.values[-1]] = del_idx
        del self.val_idx[val]

        self.values[del_idx] = self.values[-1]
        self.values.pop()
        return True
    
    def getRandom(self) -> int:
        return self.values[random.randint(0, len(self.values) - 1)]
